src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::credentials::Result;
+use crate::credentials::internal::sts_exchange::ClientAuthentication;
+use crate::headers_util::build_bearer_headers;
+use crate::token::{CachedTokenProvider, Token, TokenProvider};
+use crate::token_cache::TokenCache;
+use gax::error::CredentialsError;
+use http::{Extensions, HeaderMap};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio::time::{Duration, Instant};
+
+use super::Credentials;
+use super::dynamic::CredentialsProvider;
+use super::external_account_sources::url_sourced_account::UrlSourcedCredentials;
+use super::internal::sts_exchange::{ExchangeTokenRequest, STSHandler};
+
+#[async_trait::async_trait]
+pub(crate) trait SubjectTokenProvider: std::fmt::Debug + Send + Sync {
+    /// Generate subject token that will be used on STS exchange.
+    async fn subject_token(&self) -> Result<String>;
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSourceFormat {
+    #[serde(rename = "type")]
+    pub format_type: String,
+    pub subject_token_field_name: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSourceHeaders {
+    #[serde(flatten)]
+    pub headers: HashMap<String, String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSource {
+    pub url: Option<String>,
```

Comment:
credential source could have different fields for different BYOID types I believe.

Reply:
I was going to add more fields as we add more implementations of external accounts. I've added for now just the ones for URL sourced creds.

Reply:
You may be thinking along these lines already, but if I understand this correctly, I think this is better modeled by an enum rather than a struct:

```rust
pub enum CredentialSource {
  UrlSourced(String)
  ExecutableSource { program: String, arguments: Vec<String> }
  // ... more variants here
}
```

Reply:
I haven't thought on using enums here, how is that going to work when deserializing from JSON ? It is smart enough to detect that a given field only exists for a particular enum and translates to it ?

Reply:
> I haven't thought on using enums here, how is that going to work when deserializing from JSON ? It is smart enough to detect that a given field only exists for a particular enum and translates to it ?

It can be. See #633 for ideas, and this code too:

https://github.com/googleapis/google-cloud-rust/blob/ae91098c611434626e9809a1cab470ab81b870a0/src/gax/src/error/rpc.rs#L392-L396

Reply:
awesome, thanks for the pointers @coryan